### ISSUE TITLE
[Heroku Connect]: Filter which rows sync

### ISSUE THREAD

**Issue by selfuntitled** [heroku/roadmap#99]
### Required Terms

- [X] I agree to follow this project's [Code of Conduct](https://github.com/heroku/roadmap/blob/main/CODE_OF_CONDUCT.md)
- [X] I have read and accept the [Salesforce Program Agreement](https://www.salesforce.com/company/program-agreement/)

### What service(s) is this request for?

Connect

### Tell us about what you're trying to solve. What challenges are you facing?

```markdown
In a Salesforce Org with large data volumes, the price of using Heroku connect can get high very quickly, enough to be prohibitively expensive for many potentially reasonable use-cases. When you overlay volume based pricing for both sync and postgres storage, with the lack of sync filtering, I have no mechanism to control my Heroku Connect costs as my org grows organically, and even worse, I end up paying to sync and store data that provides me no value. In many cases, the lack of filtered sync decreases the value I get from syncing the rows I want to sync, because of the extra effort and performance impact that comes from sifting through the extra data.
```

**Comment by afawcett** [comment]
Thanks for raising this @selfuntitled . 

Curious on your thoughts/preferences on where you want to control the filtering? Within Heroku Connect or through a existing mechanism such as Sharing Rules within the attached org. Any pros/cons thoughts as to where we place such controls?

**Comment by selfuntitled** [comment]
Glad to dive deeper @afawcett  - 
Controls should be in Heroku connect - and should a standard part of the wizard to connect your org.
Sharing rules or another security based mechanism feels like it lacks discoverability on both the SF and Heroku side... from heroku, hard to guess that that's the filter mechanism, and even worse - it's rare that an accidental click on the Salesforce side could cost tens of thousands of dollars, but misconfiguring your sharing rules could very easily lead to that with HC.

**Comment by afawcett** [comment]
Thanks @selfuntitled you make a good argument here - appreciate you sharing - will revert this back into our team as we discuss future features for Heroku Connect.

**Comment by csupilow** [comment]
Came across this as we are running into this exact problem in our org, so I'd love to see it get some attention! We are using Heroku connect with an LTI tool for students, which polls Contacts (for students), Accounts (for colleges), Affiliations, and a custom object called College Application. We have 500k+ records across these objects, but the tool really only needs access to ~4% of those records overall at any given time. We purchased 250k rows believing it would provide a huge buffer, but even with prod alone were 2x over because of the lack of filtering. With a connection to full copy, we'd need to increase our (nonprofit) budget for this 4 fold. Unfortunately, it look like the inability to filter is going to price us out of using Heroku Connect until a filtering feature like this exists.

**Comment by vivekvj01** [comment]
Hi we are looking in to this. @csupilow would sharing rules an option? so you would not sync certain records.

**Comment by csupilow** [comment]
@vivekvj01 I think that probably would work in our scenario, but in full transprency this was enough of an issue that we are migrating off or Heroku Connect in the next month so others here might have more insight if they are still working on it regularly.

**Comment by vivekvj01** [comment]
@csupilow  agreed and thank you for your feedback

**Comment by mattyindustries** [comment]
Adding a vote to this.

Our Heroku use case only needs a subset of Salesforce data and the need to sync all rows massively blows out the licencing cost of Heroku

**Comment by sekhars2019** [comment]
yes, we need this feature badly. We should be able to control via some criteria or date conditions, ex- i wanted to bring only last 3 Months of data from Salesforce to Heroku. Based on last "modifieddate" I can pull 3 months of data from Salesforce to Heroku.

**Comment by nktnet1** [comment]
For anyone using sharing rules, there's an official warning in this article:

- https://help.heroku.com/7QQNUDRI/can-i-use-sharing-rules-to-restrict-record-visibility

    <img width="100%" alt="sharing-rules" src="https://github.com/user-attachments/assets/f2ecf994-1b5e-4282-b5d8-aadd6861ff37" />

    <details close>

    <summary><i>Click to view text</i></summary>

    > ## Can I use Sharing Rules to restrict record visibility?

    > ### Issue

    > You are considering Sharing Rules as a workaround to ensure Connect can only synchronize a subset of data.

    > ### Resolution

    > This approach is unsupported as it can cause mappings to become out of sync and can have an adverse impact on performance.

    > Sharing Rules don't provide any form of notification mechanism when records move in or out of scope, potentially causing them to be orphaned in your database or Salesforce. For example, Heroku Connect syncs a record that later moves out of the scope of the Sharing Rule. Connect no longer has access to it via the Salesforce APIs, but it still exists in the database.

    > Heroku Connect performs best with "View All" permissions on each mapped object. This permission removes the need for Salesforce APIs to carry out permissions checks on each record being accessed. Removing this permission from the integration user in order to use Sharing Rules can have drastic performance implications when querying and updating data in your Salesforce org.

    </details>

There really needs to be a better way to filter and sync a subset of records.

Our workaround for the time being is to use a combination of formula fields, record-triggered flows and scheduled flows to write to a custom boolean field called "is recently modified" (since sharing rules do not support formula fields or relative date).
